src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/DefaultRazorProjectHost.cs

```diff
@@ -0,0 +1,91 @@
+    // This class is responsible for intializing the Razor ProjectSnapshotManager for cases where
+    // MSBuild provides configuration support (>= 2.1).
+    [AppliesTo("DotNetCoreRazor & DotNetCoreRazorConfiguration")]
+    [Export(ExportContractNames.Scopes.UnconfiguredProject, typeof(IProjectDynamicLoadComponent))]
+    internal class DefaultRazorProjectHost : RazorProjectHostBase
+    {
+        private IDisposable _subscription;
+
+        [ImportingConstructor]
+        public DefaultRazorProjectHost(
+            IUnconfiguredProjectCommonServices commonServices,
+            [Import(typeof(VisualStudioWorkspace))] Workspace workspace)
+            : base(commonServices, workspace)
+        {
+        }
+
+        // Internal for testing
+        internal DefaultRazorProjectHost(
+            IUnconfiguredProjectCommonServices commonServices,
+             Workspace workspace,
+             ProjectSnapshotManagerBase projectManager)
+            : base(commonServices, workspace, projectManager)
+        {
+        }
+
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            await base.InitializeCoreAsync(cancellationToken).ConfigureAwait(false);
+
+            // Don't try to evaluate any properties here since the project is still loading and we require access
+            // to the UI thread to push our updates.
+            //
+            // Just subscribe and handle the notification later.
+            var receiver = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(OnProjectChanged);
+            _subscription = CommonServices.ActiveConfiguredProjectSubscription.JointRuleSource.SourceBlock.LinkTo(
+                receiver,
+                initialDataAsNew: true,
+                suppressVersionOnlyUpdates: true,
+                ruleNames: new string[] { RazorGeneral.SchemaName, });
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            await base.DisposeCoreAsync(initialized).ConfigureAwait(false);
+
+            if (initialized)
+            {
+                _subscription.Dispose();
+            }
+        }
+
+        // Internal for testing
+        internal async Task OnProjectChanged(IProjectVersionedValue<IProjectSubscriptionUpdate> update)
+        {
+            await ExecuteWithLock(async () =>
+            {
+                if (IsDisposing || IsDisposed)
+                {
+                    return;
+                }
+
+                var languageVersion = update.Value.CurrentState[RazorGeneral.SchemaName].Properties[RazorGeneral.RazorLangVersionProperty];
+
+                if (string.IsNullOrEmpty(languageVersion))
+                {
+                    // Ok we can't find a language version. Let's assume this project isn't using Razor then.
+                    await UpdateProjectUnsafeAsync(null).ConfigureAwait(false);
+                    return;
+                }
```

Comment:
I switched to this approach since it's what we'll need to do in the next step and it's just a bit simpler. It also only requires the UI thread when we need to actually push a change. The old approach did lots of hopping around.